src/core/page-parser.ts

```diff
@@ -1,67 +1,96 @@
-import {JSDOM} from "jsdom";
-import {Request} from "express";
-import {CssSelectorDefinition, CssSelectorRegistry} from "./css-selector-registry";
+import {JSDOM} from 'jsdom';
+import {snakeCase} from 'lodash';
 // @ts-ignore
 import * as RegexTranslator from 'regex-translator';
+import axios from 'axios';
 
-import {snakeCase} from 'lodash';
+import {CssSelectorDefinition, CssSelectorRegistry} from "./css-selector-registry";
 
-const axios = require('axios').default;
+export type Language = 'en' | 'de' | 'ja' | 'fr';
+export type RequestFunction = (url: string, headers?: Record<string, string>) => Promise<string>;
+
+const lodestoneBaseUrls: Record<Language, string> = {
+    'en': 'https://na.finalfantasyxiv.com/lodestone',
+    'de': 'https://de.finalfantasyxiv.com/lodestone',
+    'ja': 'https://jp.finalfantasyxiv.com/lodestone',
+    'fr': 'https://fr.finalfantasyxiv.com/lodestone',
+};
 
 export abstract class PageParser {
 
-    protected abstract getURL(req: Request): string;
+    private _language: Language;
+    protected requestFunction: RequestFunction;
+    protected baseUrl: string;
+
+    public get language() {
+        return this._language;
+    }
+
+    public set language(newLanguage: Language) {
+        this._language = Object.keys(lodestoneBaseUrls).includes(newLanguage) ? newLanguage : 'en';
+        this.baseUrl = lodestoneBaseUrls[this._language];
+    }
+
+    public constructor(language: Language = 'en', requestFunction?: RequestFunction) {
+        this._language = Object.keys(lodestoneBaseUrls).includes(language) ? language : 'en';
+        this.baseUrl = lodestoneBaseUrls[this._language];
+
+        if (typeof requestFunction === 'function') {
+            this.requestFunction = requestFunction;
+        } else {
+            const axiosInstance = axios.create();
+            this.requestFunction = (url, headers) => {
+                return axiosInstance.get(url, {headers})
+                    .then(response => response.data);
+            };
+        }
+    }
+
+    protected abstract getLodestonePage(characterId: string): Promise<string>;
 
     protected abstract getCSSSelectors(): CssSelectorRegistry;
 
-    public async parse(req: Request, columnsPrefix = ''): Promise<Object> {
-        const {data} = await axios.get(this.getURL(req)).catch((err: any) => {
-            throw new Error(err.response.status);
+    protected modifyParseResult(data: Record<string, unknown>): Record<string, unknown> {
+        return data;
+    }
+
+    public async get(characterId: string, columns: string[] = []): Promise<Record<string, unknown>> {
```

Comment:
I'm not sure making it `get` with only one parameter is the way to go for the API. While it serves the purpose for character data, having an ID as parameter makes no sense for the search endpoints, naming it characterId also makes it "character-oriented" which is not the goal of the PageParser class, since it's here to carry all the parsing logic for anyone to use.

Maybe this means that we should have an ` abstract class CharacterPareParser extends PageParser` that provides common character logic without making `PageParser` too character oriented.

Reply:
I was too narrow with arguments of the `PageParser` here, since I did not see the full scope of the project yet, sorry!
I was too focused on the character/profile pages, since that was what I needed for my use case.

As discussed on Discord already, we should probably use an options object to provide more flexibility for the subclasses.
I also agree with your idea of specializing a "character page parser" class and will update the PR accordingly.
Since the "character" pages are called "profile" in the lodestone-css-selectors, do we want to name it `ProfilePageParser`?

Reply:
With the additions in #6 , I don't think a Character parser abstract class is needed anymore, since in the end, it's just one parser per request, and everything about a character is handled by the same parser, except minions and mounts but they'll work fine with the new options-based API 👍